sdks/python/tests/unit/test_opik_configure.py

```diff
@@ -1,97 +1,744 @@
+        Test that the URL is successfully returned on the second attempt after the first failure.
+        """
+        result = _ask_for_url()
+        assert result == "http://valid-url.com"
+        assert mock_is_instance_active.call_count == 2
+
+    @patch(
+        "builtins.input",
+        side_effect=[
+            "http://invalid-url.com",
+            "http://invalid-url-2.com",
+            "http://valid-url.com",
+        ],
+    )
+    @patch("opik.opik_configure.is_instance_active", side_effect=[False, False, True])
+    def test_ask_for_url_success_on_third_try(
+        self, mock_is_instance_active, mock_input
+    ):
+        """
+        Test that the URL is successfully returned on the third attempt after two failures.
+        """
+        result = _ask_for_url()
+        assert result == "http://valid-url.com"
+        assert mock_is_instance_active.call_count == 3
+
+    @patch("builtins.input", side_effect=["http://invalid-url.com"] * 3)
+    @patch("opik.opik_configure.is_instance_active", return_value=False)
+    @patch("opik.opik_configure.LOGGER.error")
+    def test_ask_for_url_logging(
+        self, mock_logger_error, mock_is_instance_active, mock_input
+    ):
+        """
+        Test that errors are logged when the URL is not accessible.
+        """
         with pytest.raises(ConfigurationError):
-            configure(
-                api_key=api_key,
-                url=url,
-                workspace=workspace,
-                force=True,
-                use_local=local,
-            )
-    else:
-        # No exception should be raised
-        configure(
-            api_key=api_key, url=url, workspace=workspace, force=True, use_local=local
+            _ask_for_url()
+
+        assert mock_logger_error.call_count == 3
+        mock_logger_error.assert_called_with(
+            f"Opik is not accessible at http://invalid-url.com. Please try again, the URL should follow a format similar to {OPIK_BASE_URL_LOCAL}"
+        )
+
+
+class TestAskForApiKey:
+    @patch("opik.opik_configure.getpass.getpass", return_value="valid_api_key")
+    @patch("opik.opik_configure.is_api_key_correct", return_value=True)
+    def test_ask_for_api_key_success(self, mock_is_api_key_correct, mock_getpass):
+        """
+        Test successful entry of a valid API key.
+        """
+        result = _ask_for_api_key()
+        assert result == "valid_api_key"
+        mock_is_api_key_correct.assert_called_once_with("valid_api_key")
+
+    @patch("opik.opik_configure.getpass.getpass", return_value="invalid_api_key")
+    @patch("opik.opik_configure.is_api_key_correct", return_value=False)
+    def test_ask_for_api_key_all_retries_fail(
+        self, mock_is_api_key_correct, mock_getpass
+    ):
+        """
+        Test that after 3 invalid API key attempts, a ConfigurationError is raised.
+        """
+        with pytest.raises(ConfigurationError, match="API key is incorrect."):
+            _ask_for_api_key()
```

Comment:
`_ask_for_api_key()` is a private function. If someone sees such functions, they might want to rename them, merge with another function - and we don't want the test to file.
If it's private - then it's an implementation detail and shouldn't be tested.
If it's a testable unit of functionality - it should be moved to a helper module.